### PR TITLE
Add minimal NoteSubFolderApi implementation and ScriptingService methods

### PR DESCRIPTION
**Pull request by fnkabit**
This would be required in order to implement a script that exports all the notes in the active folder ( feature #1165 )

### CODE REVIEW

**Comment by fnkabit** [comment]
@pbek Something wrong with the build here. `moc` is not run on `NoteSubFolderApi` for some reason ?

File: src/QOwnNotes.pro [hunk 1]
```diff
@@ -146,6 +146,7 @@ SOURCES += main.cpp\
     widgets/navigationwidget.cpp \
     widgets/notepreviewwidget.cpp \
     api/noteapi.cpp \
+    api/notesubfolderapi.cpp \
```
Comment (pbek) [inline src/QOwnNotes.pro:149]: Cab you please add these also to the CMakeFile

File: src/api/notesubfolderapi.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+NoteSubFolderApi *NoteSubFolderApi::fromNoteSubFolder(const NoteSubFolder &noteSubFolder) {
+    auto* noteSubFolderApi = new NoteSubFolderApi();
+    noteSubFolderApi->fetch(noteSubFolder.getId());
+    return noteSubFolderApi;
+}
+
+/**
+ * Returns all linked notes
```
Comment (pbek) [inline src/api/notesubfolderapi.cpp:27]: Can you please use `Returns all notes of the note subfolder".

File: src/services/scriptingservice.cpp [hunk 3]
```diff
@@ -54,6 +55,7 @@ ScriptingService::ScriptingService(QObject *parent) : QObject(parent) {
     qmlRegisterType<TagApi>("com.qownnotes.tagapi", 1, 0, "TagApi");
 
     qmlRegisterType<NoteApi>("QOwnNotesTypes", 1, 0, "Note");
+    qmlRegisterType<NoteSubFolderApi>("QOwnNotesTypes", 1, 0, "NoteSubFolder");
```
Comment (pbek) [inline src/services/scriptingservice.cpp:58]: Can you please also add documentation on https://www.qownnotes.org/scripting/classes.html.

File: src/services/scriptingservice.cpp [hunk 4]
```diff
@@ -2012,6 +2014,27 @@ QVariant ScriptingService::getApplicationSettingsVariable(
     return settings.value(key, defaultValue);
 }
 
+/**
+ * Fetches note subfolders that are children of parentId
+ *
+ * @param parentId int the id of the parent note subfolder
+ * @return QList<QObject*>
+ */
+QList<QObject*> ScriptingService::fetchNoteSubFoldersByParentId(int parentId)
```
Comment (pbek) [inline src/services/scriptingservice.cpp:2023]: Can you please also add documentation.

Comment (pbek) [reply src/services/scriptingservice.cpp:2023]: Wouldn't it be better to make this method part of `api/notesubfolderapi.cpp`.

File: src/services/scriptingservice.cpp [hunk 5]
```diff
@@ -2150,6 +2173,22 @@ bool ScriptingService::fileExists(const QString &filePath) const {
     return file.exists();
 }
 
+/**
+ * Fetches a note subfolder by its id
+ *
+ * @param id int the id of the note subfolder
+ * @return NoteSubFolderApi*
+ */
+NoteSubFolderApi *ScriptingService::fetchNoteSubFolderById(int id)
```
Comment (pbek) [inline src/services/scriptingservice.cpp:2182]: Wouldn't it be better to make this method part of `api/notesubfolderapi.cpp`.

**Comment by pbek** [comment]
> @pbek Something wrong with the build here. `moc` is not run on `NoteSubFolderApi` for some reason ?

Are you talking about the tests? There are different project file for those. You might add some includes there too.

**Comment by fnkabit** [comment]
> > @pbek Something wrong with the build here. `moc` is not run on `NoteSubFolderApi` for some reason ?
> 
> Are you talking about the tests? There are different project files for those. You might need to add some includes there too.

No, I didn't know those builds were using `CMakeLists.txt`; I thought the .pro is used and didn't know why I had those undefined reference errors.

**Comment by pbek** [comment]
> No, I didn't know those builds were using CMakeLists.txt

The builds here on GitHub don't use cmake, only qmake. But other builds processes do...

File: src/services/scriptingservice.cpp (hunk 3 above)
Comment (fnkabit) [reply src/services/scriptingservice.cpp:58]: Documentation: PR #1931

**Pull request merged**